tests/post/steps/test_liveness.py

```diff
@@ -34,58 +35,76 @@ def check_resource_list(host, resource, namespace):
 
 
 @then(parsers.parse(
-    "we can exec '{command}' in the "
-    "'{pod}' pod in the '{namespace}' namespace"))
-def check_exec(host, command, pod, namespace):
+    "we can exec '{command}' in the pod labeled '{label}' "
+    "in the '{namespace}' namespace"))
+def check_exec(host, command, label, namespace):
+    candidates = _get_pods(host, label, namespace)
+    try:
+        pod, = candidates
+    except ValueError:
+        pytest.fail(
+            "Expected one (and only one) pod with label {l}, found {f}".format(
+                l=label, f=len(candidates)
+            )
+        )
+
     cmd = ' '.join([
         'kubectl',
         '--kubeconfig=/etc/kubernetes/admin.conf',
         'exec',
         '--namespace {0}'.format(namespace),
-        pod,
+        pod['metadata']['name'],
         command,
     ])
 
     with host.sudo():
         host.check_output(cmd)
 
 
-@then(
-    parsers.parse(
-        "we have at least {min_pods_count:d} running pod labeled '{label}'"
-    )
-)
+@then(parsers.parse(
+    "we have at least {min_pods_count:d} running pod labeled '{label}'"))
 def count_running_pods(host, min_pods_count, label):
-    cmd = (
-        'kubectl --kubeconfig=/etc/kubernetes/admin.conf '
-        'get pod -l "{label}" --field-selector=status.phase=Running '
-        '--namespace kube-system -o json'
-    ).format(label=label)
-
     # Just after the deployment, the pods may not in running state yet
     # so implement a timeout
     interval = 3
     attempts = 10
 
-    with host.sudo():
-        for _ in range(attempts):
-
-            ret = host.run(cmd)
-            assert ret.rc == 0, ret.stdout
+    for _ in range(attempts):
+        pods = _get_pods(
+            host,
+            label,
+            namespace="kube-system",
+            status_phase="Running",
+        )
 
-            pods = json.loads(ret.stdout)
+        if len(pods) < min_pods_count:
+            time.sleep(interval)
+        else:
+            # if nb pods is >= then it's ok
+            break
+
+    else:
+        assert len(pods) >= min_pods_count, (
+            "Expected at least {e} running pods labeled with {l} but "
+            "found only {f} after waiting {t} seconds"
+        ).format(
+            e=min_pods_count,
+            f=len(pods),
+            l=label,
+            t=str(interval * attempts)
+        )
+
+
+# Utilities
+def _get_pods(host, label, namespace='default', status_phase='Running'):
+    cmd = (
+        'kubectl --kubeconfig=/etc/kubernetes/admin.conf '
+        'get pods -l "{label}" --field-selector=status.phase={status_phase} '
+        '--namespace {namespace} -o json'
+    ).format(label=label, namespace=namespace, status_phase=status_phase)
 
-            if len(pods['items']) < min_pods_count:
-                time.sleep(interval)
-            else:
-                # if nb pods is >= then it's ok
-                break
+    with host.sudo():
+        result = host.run(cmd)
```

Comment:
How does that `run` method works?

If it's like `Popen` then it's better to pass a list of string instead of a string (since you would need a shell, and thus escape the parameters).

Reply:
It's not accepting a list of strings, but it handles quoting the parameters to guard against shell injection, if that's your concern.

The method is documented [here](https://testinfra.readthedocs.io/en/latest/modules.html#testinfra.host.Host.run).
It seems to first [`quote` the string](https://github.com/philpep/testinfra/blob/bbab44504eec8d5c8995e3473166f6e57f9d88f9/testinfra/backend/base.py#L151-L155), which is used to format the `command` with `pipes.quote`d positional arguments.

Let's keep it like this for now and see if we encounter issues with it in the future.

Reply:
IIUC, in our case it's not quoted as we pass directly a whole string (but the method does escape params if called properly).

Reply:
Right, I'll improve it in the following PR